### PR TITLE
[Event Hubs Client] Event Processor<T> Main Loop

### PR DESCRIPTION
**Pull request by jsquire**
# Summary

The focus of these changes is to implement the main loop of the processor that executes in the background when the processor is running, and is responsible for coordinating with the load balancer for partition ownership changes and acting as a sentinel for the active partition processors.

# Last Upstream Rebase

Saturday, March 21, 11:58am (EST)

# References and Related Issues 

- [.NET Event Hubs Client: Event Processor<T> Proposal](https://gist.github.com/jsquire/1f4a1e72fc02871f443ce365222d40f6)
- [Implement the Event Processor Base](https://github.com/Azure/azure-sdk-for-net/issues/9324) (#9324)

### CODE REVIEW

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Core/ObjectExtensions.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Core/ObjectExtensions.cs]: For context:  We were starting to build up several negative event filters, sometimes with multiple clauses, that looked like:

```csharp
try 
{
    ...
}
catch (Exception ex) when (!(ex is SomeType))
{
    ...
}
catch (Exception ex) when ((!(ex is OtherType) && (!(ex is ThirdType)))
{
    ...
}
```

With this extension, they transform into a form which I find easier to read and parse:

```csharp
try 
{
    ...
}
catch (Exception ex) when (ex.IsNotType<SomeType>())
{
    ...
}
catch (Exception ex) when ((ex.IsNotType<OtherType>()) && (ex.IsNotType<ThirdType>()))
{
    ...
}
```

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs]: This outer try/catch is intended to serve as a catch-all for scenarios that may otherwise crash the processor without invoking the error handler.   For example, the current processor would silently fail to start if creating a connection failed.  Outside of unexpected scenarios, it isn't intended that we bubble up to the outer level.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs]: For some reason, using `string.IsNullOrEmpty(partitionId)` here would not trigger this path when it was `null`.    I had to use a local capture here with an explicit `""` as a result.  I'd be interested in any insight that folks have on why that may have been the case.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs]: Full test coverage for the blocks on 1162-1165 and 1173 - 1188 fit more naturally with the Start/Stop partition context, so I've deferred them for now.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs]: This was needed in order to avoid a potential deadlock when using the synchronous `StopProcessing` call within the error handler, which is a case that we want to support.

File: sdk/eventhub/Azure.Messaging.EventHubs/tests/Primitives/EventProcessorTests.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/tests/Primitives/EventProcessorTests.cs]: The timing for this was tricky; even without the safe-guard loop at the end, it still adds roughly 4 seconds to the local run. _(from ~8.5 seconds to ~12.4 seconds)_    That said, it proved valuable as it caught a deadlock in what should be a valid scenario, so I definitely do not want to remove it.

I'd be grateful for any thoughts on how to potentially tune it or better approaches that allow for deterministic validation.

File: sdk/eventhub/Azure.Messaging.EventHubs/tests/Primitives/EventProcessorTests.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/tests/Primitives/EventProcessorTests.cs]: These are intentionally a longer duration than the cancellation token timeout; my intent is to prove for each of these scenarios that the error handler does not block other operations.   I considered covering that in an additional test, but there are a non-trivial number of cases and this isn't a vital core scenario that I believe requires a specific dedicated test.

**Comment by jsquire** [comment]
//cc: @christothes

File: sdk/eventhub/Azure.Messaging.EventHubs/tests/Primitives/EventProcessorTests.cs
Comment (kinelski) [inline sdk/eventhub/Azure.Messaging.EventHubs/tests/Primitives/EventProcessorTests.cs]: Not sure if intentional, but single-line try/catch blocks like this in this file have two blank spaces before the `catch` keyword. Just mentioning in case the extra space was an accident.

File: sdk/eventhub/Azure.Messaging.EventHubs/tests/Primitives/EventProcessorTests.cs
Comment (kinelski) [inline sdk/eventhub/Azure.Messaging.EventHubs/tests/Primitives/EventProcessorTests.cs]: There are a few test names with `Processing` misspelled as `Processin`.

File: sdk/eventhub/Azure.Messaging.EventHubs/tests/Primitives/EventProcessorTests.cs
Comment (kinelski) [reply sdk/eventhub/Azure.Messaging.EventHubs/tests/Primitives/EventProcessorTests.cs]: Instead of setting the result of the `completionSource` in the `StopProcessing` callback, couldn't we set it when calling `OnProcessingErrorAsync`, after the try/catch block? This way we can make sure the processor has stopped before the verification. I don't think this will be very helpful in improving the total time, though.

**Review by kinelski — APPROVED** [review]
LGTM

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs
Comment (ellismg) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs]: I guess this would have been impossible if T had a constraint on it being an exception type (which is what the documentation for the helper implied).  Maybe add that constraint to prevent stuff like this?

**Review by ellismg — CHANGES_REQUESTED** [review]
A few questions from me.  I think the only blocking feedback is about the use of `ex.IsNotType` on L1443 of `EventProcessor{TPartition}.cs`

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Core/ObjectExtensions.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Core/ObjectExtensions.cs]: Yeah, that `typeparam` text is a mistake on my part.   I started off with this scoped to `Exception` types with constraints and then realized that was a silly limitation and generalized.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs]: I'm a big fan of switch expressions/pattern matching.   While I acknowledge that it still looks weird to many, since it was only recently introduced in C# 8, but I expect that we'll see that getting more familiar going forward.  Many of the same concerns were raised around the use of pattern matching when it was first introduced, for example, and all the way back to generics.

Since we're fully embracing the C# 8 features within the codebase, including the use of switch expressions elsewhere, I'd prefer to keep it as is.  That said, I'm open if you feel strongly about it, though we'll probably want to extract out to a helper method in that case to avoid the `catch` block getting too chunky.  _(it's already borderline, in my opinion)_

File: sdk/eventhub/Azure.Messaging.EventHubs/tests/Primitives/EventProcessorTests.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/tests/Primitives/EventProcessorTests.cs]: I could have sworn that I had it that way at one point and moved it over for some reason.   That said, it does make sense to have it fire after `OnProcessingErrorAsync` has returned, since that *should* ensure that the Stop call has fully completed.    Looks like the execution time remains the same, sadly.

I'm going to leave the safety loop in there just to guard against non-determinism, since the `Status` property intentionally does not attempt to acquire the semaphore when the processor is stopping, there is still a small (but acceptably small) chance that it could see `Stopping` rather than `Stopped` due grabbing a copy of the override value at exactly the wrong moment.

File: sdk/eventhub/Azure.Messaging.EventHubs/tests/Primitives/EventProcessorTests.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/tests/Primitives/EventProcessorTests.cs]: You're right, of course, and I was being lazy because it cascades into a few other tests.  I've been shamed into doing the right thing.   😄

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs]: Chris moved a few things around so that we're reading from the `Data` property bag on the exception and not making assumptions about the message, which is intended as free-form text.   That's also the vehicle for sending the partition context along now.

File: sdk/eventhub/Azure.Messaging.EventHubs/tests/Primitives/EventProcessorTests.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/tests/Primitives/EventProcessorTests.cs]: It's intentional; I like the extra space when doing this kind of "ignore all the things" just to give some visual distance to the catch block.   I should consider doing an extension for `Task` in the next round for something like `SafeAwait` or a helper method or something.    I'll give it some thought and defer to the next round.

File: sdk/eventhub/Azure.Messaging.EventHubs/tests/Primitives/EventProcessorTests.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/tests/Primitives/EventProcessorTests.cs]: Would you believe that I wrote this if there weren't at least a couple of typos sprinkled in?   😋

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Core/ObjectExtensions.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Core/ObjectExtensions.cs]: Yeah, that's a mistake on my part.  I started with this constrained to only the `Exception` family and then opened it up.   Fixed up.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs]: In theory, we shouldn't see an `OperationCanceledException` instance here because we wrap them into a `TaskCanceledException` for known cancellations, so the only ones that should get to this level are most likely an indication that something went wrong at the service communication level within the AMQP library that its built on.   

For a bit wider context, the [service documentation](https://docs.microsoft.com/en-us/azure/event-hubs/event-hubs-messaging-exceptions) calls out that operation cancelled indicates that somewhere you closed or disposed something.  Because of this, we're trying to avoid confusion when normal and expected cancellation scenarios happen by consistently surfacing only `TaskCanceledException` rather than being agnostic to both.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs]: This is intentional; the service team is working on a change for "some time in the near future" that will allow dynamic increases to partitions.  While your suggestion would work just fine today, this gives us a bit of future proofing.

Since we're not in a hot path with respect to impacting throughput for processing partitions which are already owned and there's an inherent latency with the load balancer cycles, I think the trade-off is worth it, especially since the query is made over an existing connection/link to the service and doesn't have the overhead of the HTTP handshake to consider.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs]: I don't feel strongly, as we're using well-known members of the `Data` bag through dedicated extension methods, so the fields aren't being misused nor are we parsing something like the message.  I think the history here started before the load balancer was a self-contained type and the processor client was coming in very hot for a release due to non-trivial churn late in the game.

Personally, I'd probably consider an internal exception type specific to the load balancer scenario, similar to what we did with the `DeveloperCodeException`.  Though, I'm going to defer this for now; there are some upcoming enhancements to load balancing that are likely to require changing this flow and I would prefer not to bring the load balancer into the scope of these changes.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs]: The first and second are far more likely; the third is possible but unlikely.   Here's the breakdown, as I see it:

- If the load balancer encounters an error that isn't specific to a partition, then we would hit the first case.  This could be in querying the storage provider for a list of ownerships or similar.

- If the load balancer encounters an error renewing ownership for a partition that is owned and which we're currently processing, we'd hit the second case.  This could be in requesting the storage provider update the ownership records to extend a partition that we already own.

- If the load balancer is making a decision about a partition that we would like to claim and believes that it is owned, but processing hasn't fully started yet, or the attempt to start processing the partition failed, then we would hit the third case.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs]: gah.   That's... not awesome.  I had started with that extension constrained to only `Exception` types, but then opened it to all the things, as I can see it being useful in other contexts, albeit marginally.

Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs]: I'm going to leave it as-is for now, but I took a note to give it a look during the next round of changes and consider scoping it back to only exceptions.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs]: That's a good thought.  It's a bit late to push into this revision, but I'll play with it tomorrow with fresh eyes and see how it looks pulled out.

File: sdk/eventhub/Azure.Messaging.EventHubs/tests/Primitives/EventProcessorTests.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/tests/Primitives/EventProcessorTests.cs]: There's actually significance to the ordering.   We don't want the callback to run until after the connection has been returned, which ensures that we're in the processing loop and running.   

The load balancer setup isn't as fussy in this case; we could move forward when the call was made or after it returned.    I swapped the ordering on the load balancer, which makes them consistent and eases my OCD (which you seem to share.  :smile:)

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs
Comment (ellismg) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs]: > This is intentional; the service team is looking to allow dynamically increasing the partitions at some point in the not-too-distant future.

Gotcha.  Thanks!

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs
Comment (ellismg) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs]: SGTM

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs
Comment (ellismg) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs]: SGTM

**Comment by jsquire** [comment]
Blocking issues and formatting have been addressed; there's a bit of restructuring that @ellismg suggested that I'm going to kick the tires on before merging this in.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs
Comment (ellismg) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs]: > If the load balancer is making a decision about a partition that we would like to claim and believes that it is owned, but processing hasn't fully started yet, or the attempt to start processing the partition failed, then we would hit the third case.

In this case, do you think there is value in passing as "partially initialized" `TPartition` in, vs say `null`?  I get that we're hiding a small amount of information here but I wonder if that leads to a nicer overall model.

Just something to noodle over.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs
Comment (ellismg) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs]: Don't feel super strongly.

**Review by ellismg — APPROVED** [review]
LGTM

File: sdk/eventhub/Azure.Messaging.EventHubs/tests/Primitives/EventProcessorTests.cs
Comment (christothes) [reply sdk/eventhub/Azure.Messaging.EventHubs/tests/Primitives/EventProcessorTests.cs]: Random idea -- If you provide a mock implementation for `Logger` the completionSource could be set on the callback to `EventProcessorStopComplete` instead, which happens after the `Status` is stable.

**Review by christothes — COMMENTED** [review]
LGTM - not much to add here other than a couple minor thoughts.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Core/ObjectExtensions.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Core/ObjectExtensions.cs]: ... and it's now back with the change to scope the extension method only to exception scenarios.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs]: It's a fair question, and I don't know that I have a solid answer.  Passing the partially initialized object does allow the partition identifier to be communicated, which may offer insight into partition-specific issues or help diagnose that there is an issue with the load balancer putting too much contention on ownership for a single partition.

On the other hand, you're not wrong in that it is potentially confusing and may not provide the full set of context for a custom type.   I'm kind of banking on somewhat of an understanding given that the initialize handler won't have fired for the partition and that is where developers would perform the initialization of the custom type. 

I'm certainly open to further discussion and better ideas.   In the meantime, I'll add some additional comments to the error handler to try and clarify that the partition may only be minimally initialized when called.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs]: I lied; it ended up annoying me enough that I made the change now.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs]: Thanks... this turned out to be a great idea.  I think that it improves the focus quite a bit.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs]: ooohh... I like that.  Done!

File: sdk/eventhub/Azure.Messaging.EventHubs/tests/Primitives/EventProcessorTests.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/tests/Primitives/EventProcessorTests.cs]: Clever idea!  Gave it a try and it looks stable and deterministic in local runs without the safety loop.   Bringing it forward into these changes and will adjust if we see anything weird in nightly or CI runs.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs
Comment (ellismg) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs]: > I'll add some additional comments to the error handler to try and clarify that the partition may only be minimally initialized when called.

Thanks!

**Pull request merged**
